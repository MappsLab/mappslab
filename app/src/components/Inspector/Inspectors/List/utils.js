// @flow
import type { ClassroomType, UserType, MapType } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<NodeArray> | NodeArray

export type ListItemHandler = (ClassroomType | UserType | MapType) => void

export type ListItem = {
	node: ClassroomType | UserType | MapType,
	key: string,
	onClick: () => void,
}

// const nodeToInspectorItem = (item: ClassroomType | UserType | MapType): InspectorItem => ({
// 	uid: item.uid,
// 	type: item.__typename,
// 	title: item.__typename === 'User' ? item.name : item.title,
// })

export const nodeToListItem = (node: ClassroomType | UserType | MapType, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
