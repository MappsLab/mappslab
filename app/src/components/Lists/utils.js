// @flow
import type { Node, Variables } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

// type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<void> | void

export type ListItemHandler = (Node) => void

export type ListItem = {
	node: Node,
	onClick: () => void,
}

export type ListOfTypeBaseProps<Type> = {
	title: string,
	items: Array<Type>,
	viewerCanAdd: boolean,
	update: (Type) => void,
	onItemClick: ListItemHandler,
}

export type ListOfTypeProps<Type> = ListOfTypeBaseProps<Type> & {
	searchQuery: (Variables) => Promise<void>,
	searchResults: Array<Type>,
}

export const nodeToListItem = (node: Node, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
