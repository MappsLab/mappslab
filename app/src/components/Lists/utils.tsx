import { NodeType } from '../../types-ts'

export type SearchForList = (arg0: any) => Promise<void> | void

export type ListItemHandler<T extends NodeType> = (
	item: T,
) => void | Promise<void>

export type CreateNewFn = (str: string) => void | Promise<void>

export interface ListItem {
	key: string
	node: NodeType
	info: Array<string>
	onClick: () => void | Promise<void>
}

export interface ListOfTypeProps<Type extends NodeType> {
	title: string
	items?: Type[]
	viewerCanAdd: boolean
	update?: (node: NodeType) => void | Promise<void>
	onItemClick?: ListItemHandler<Type>
	onSearchResultClick?: ListItemHandler<Type>
	create?: (args: any) => Promise<void> | void
	remove?: (args: any) => Promise<void> | void
}

export const nodeToListItem = <T extends NodeType>(
	node: T,
	handler: ListItemHandler<T>,
): ListItem => {
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
