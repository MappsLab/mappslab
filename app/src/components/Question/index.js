// @flow

export { QuestionProvider, QuestionConsumer } from './QuestionProvider'
export { default as QuestionDialog } from './QuestionDialog'
