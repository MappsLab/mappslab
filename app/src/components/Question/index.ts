
export * from './QuestionProvider'
export * from './QuestionDialog'
