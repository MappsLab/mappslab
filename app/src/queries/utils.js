// @flow
import * as R from 'ramda'
import type { LoadingState } from 'Types/GraphQL'

type Node = {
	cursor: string,
	node: any,
}

type withEdges = {
	edges: Array<Node>,
	pageInfo?: {
		hasNextPage: boolean,
		hasPrevPage: boolean,
		cursor: string,
	},
}

const splitEdges = (obj: withEdges) => ({
	edges: obj.edges.map(({ cursor, node }) => ({ __cursor: cursor, ...node })),
	pageInfo: obj.pageInfo || {},
})

// export const unwindEdges = (obj: any) =>
// 	typeof obj === 'object'
// 		? Object.entries(obj).reduce((acc, [key, value]) => {
// 				if (value && value.edges && Array.isArray(value.edges)) {
// 					// $FlowFixMe
// 					const { edges, pageInfo } = splitEdges(value)
// 					const pageInfoKey = `${key}PageInfo`
// 					return {
// 						[key]: edges.map(unwindEdges),
// 						[pageInfoKey]: pageInfo,
// 						...acc,
// 					}
// 				}
// 				const parsed =
// 					typeof value === 'object'
// 						? // unwind recursively if there are more values
// 						  Array.isArray(value)
// 							? // if it's an array, unwind each item as a new array
// 							  value.map(unwindEdges)
// 							: // otherwise, pass in the object as is
// 							  unwindEdges(value)
// 						: // otherwise, value is OK
// 						  value
// 				return {
// 					[key]: parsed,
// 					...acc,
// 				}
// 		  }, {})
// 		: obj

export const unwindEdges = (o: Object): Object => {
	if (typeof o !== 'object') return o
	return R.pipe(
		R.toPairs,
		// Iterate over the properties and their values
		R.reduce(
			(acc, [key, value]) =>
				value && value.edges && Array.isArray(value.edges)
					? // When the value has an 'edges' property that is an array,
					  R.pipe(
							// Pull out the 'pageInfo' prop and rename it
							R.assoc(`${key.replace(/Connection$/, '')}PageInfo`, value.pageInfo || {}),
							// And pluck out the nodes
							R.assoc(key.replace(/Connection$/, ''), R.pluck('node', value.edges).map(unwindEdges)),
					  )(acc)
					: // otherwise, if it's an object (and not null bc null is an object)
					  value && typeof value === 'object'
						? // unwind it
						  R.assoc(key, Array.isArray(value) ? value.map(unwindEdges) : unwindEdges(value))(acc)
						: // lastly, return the value as is
						  R.assoc(key, value)(acc),
			{},
		),
	)(o)
}

export const getNetworkStatus = (num: number): LoadingState => {
	switch (num) {
		case 1:
			return 'loading'
		case 4:
			return 'refetching'
		case 2:
		case 6:
			return 'passivelyRefetching'
		case 3:
			return 'fetchingMore'
		case 7:
			return 'ready'
		case 8:
			return 'errors'
		default:
			throw new Error(`Status number ${num} is not a valid network status`)
	}
}
