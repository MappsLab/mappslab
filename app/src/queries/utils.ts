import * as R from 'ramda'

type LoadingState =
	| 'loading'
	| 'refetching'
	| 'passivelyRefetching'
	| 'fetchingMore'
	| 'ready'
	| 'errors'

type Node = {
	cursor: string
	node: any
}

type withEdges = {
	edges: Array<Node>
	pageInfo?: {
		hasNextPage: boolean
		hasPrevPage: boolean
		cursor: string
	}
}

const splitEdges = (obj: withEdges) => ({
	edges: obj.edges.map(({ cursor, node }) => ({ __cursor: cursor, ...node })),
	pageInfo: obj.pageInfo || {},
})

export const getNetworkStatus = (num: number | void): LoadingState => {
	switch (num) {
		case undefined:
		case 1:
			return 'loading'
		case 4:
			return 'refetching'
		case 2:
		case 6:
			return 'passivelyRefetching'
		case 3:
			return 'fetchingMore'
		case 7:
			return 'ready'
		case 8:
			return 'errors'
		default:
			throw new Error(`Status number ${num} is not a valid network status`)
	}
}
