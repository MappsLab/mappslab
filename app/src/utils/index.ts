export * from './data'
export * from './listeners'
export * from './maps'
export * from './media'
export * from './statecharts'
export * from './storage'
export * from './url'
export * from './parsing'
