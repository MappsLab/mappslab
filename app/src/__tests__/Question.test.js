// @flow
/* eslint-disable no-undef */
import * as React from 'react'
import { fireEvent, wait } from 'react-testing-library'
import { render } from 'Jest/utils'
import { QuestionProvider, QuestionConsumer, QuestionDialog } from 'Components/Question'
import type { QuestionContext } from 'Components/Question/QuestionProvider'

const QuestionWrapper = ({ children }: { children: (QuestionContext) => React.Node }) => (
	<QuestionProvider>
		<QuestionConsumer>{children}</QuestionConsumer>
	</QuestionProvider>
)

const BasicQuestion = () => (
	<QuestionWrapper>
		{({ ask }: QuestionContext) => (
			<React.Fragment>
				<QuestionDialog />
				<button
					type="button"
					data-testid="button"
					onClick={() => {
						ask({ message: 'are you sure?', options: [{ title: 'YES' }, { title: 'NO', level: 'secondary' }] })
					}}
				>
					click?
				</button>
			</React.Fragment>
		)}
	</QuestionWrapper>
)

describe('Question Component', () => {
	it('Mounts Correctly', () => {
		const { container } = render(<BasicQuestion />)
		expect(container).toMatchSnapshot()
	})

	it('should render its children', async () => {
		const { getByTestId } = render(<BasicQuestion />)
		expect(getByTestId('button')).toHaveTextContent('click?')
	})

	it('should not render a <QuestionDialog> by default', async () => {
		const { queryByTestId } = render(<BasicQuestion />)
		expect(queryByTestId('alert')).toBeFalsy()
	})

	it('should render an <QuestionDialog> when a consumer calls `ask`', async () => {
		const { queryByTestId, queryByText, getByTestId } = render(<BasicQuestion />)
		const button = getByTestId('button')
		expect(queryByTestId('alert')).toBeFalsy()
		fireEvent.click(button)
		expect(queryByTestId('alert')).toBeTruthy()
		expect(queryByText('are you sure?')).toBeTruthy()
		expect(queryByText('YES')).toBeTruthy()
		expect(queryByText('NO')).toBeTruthy()
	})

	it('should return the value provided by the config when `answerQuestion` is called on an option', async () => {
		const reportAnswer = jest.fn()
		const { getByTestId } = render(
			<QuestionWrapper>
				{({ ask }: QuestionContext) => (
					<React.Fragment>
						<QuestionConsumer>
							{({ currentQuestion }) => {
								if (!currentQuestion || !currentQuestion.options) return null
								return (
									<React.Fragment>
										{currentQuestion.options.map(({ answerQuestion, title, testId }) => (
											/* Here we are also testing that custom props like testId can be passed through */
											<button key={testId} type="button" data-testid={testId} onClick={answerQuestion}>
												{title}
											</button>
										))}
									</React.Fragment>
								)
							}}
						</QuestionConsumer>
						<button
							type="button"
							data-testid="askButton"
							onClick={async () => {
								const answer = await ask({
									message: 'are you sure?',
									options: [
										{ title: 'YES', testId: 'yah', returnValue: true },
										{ title: 'NO', testId: 'nah', returnValue: false },
									],
								})
								reportAnswer(answer)
							}}
						>
							click?
						</button>
					</React.Fragment>
				)}
			</QuestionWrapper>,
		)
		const askButton = getByTestId('askButton')
		/* open the dialog */
		fireEvent.click(askButton)

		/* answer yah */
		const yahButton = getByTestId('yah')
		fireEvent.click(yahButton)
		await wait()

		/* open the dialog */
		fireEvent.click(askButton)

		/* answer nah */
		const nahButton = getByTestId('nah')
		fireEvent.click(nahButton)
		await wait()
		expect(reportAnswer.mock.calls).toHaveLength(2)
		expect(reportAnswer.mock.calls[0][0]).toEqual(true)
		expect(reportAnswer.mock.calls[1][0]).toEqual(false)
	})

	/* TO DO */
	it('should use a custom `answerWith` function to transform the return value', async () => {
		/* Arrange */
		const reportAnswer = jest.fn()
		const getGithubInfo = jest.fn((handle) => `https://github.com/${handle}`)
		const { getByTestId } = render(
			<QuestionWrapper>
				{({ ask, answered }: QuestionContext) => (
					<React.Fragment>
						<QuestionConsumer>
							{({ currentQuestion }) => {
								if (!currentQuestion || !currentQuestion.options) return null
								return (
									<React.Fragment>
										{currentQuestion.options.map(({ answerQuestion, title, testId }) => (
											/* Here we are also testing that custom props like testId can be passed through */
											<button key={testId} type="button" data-testid={testId} onClick={answerQuestion}>
												{title}
											</button>
										))}
									</React.Fragment>
								)
							}}
						</QuestionConsumer>
						<button
							type="button"
							data-testid="askButton"
							onClick={async () => {
								const answer = await ask({
									message: 'Pick a profile',
									options: [
										{ title: 'Good Idea', testId: 'answerBtn1', returnValue: 'good-idea', answerWith: getGithubInfo },
										{ title: 'Bad Idea', testId: 'answerBtn2', returnValue: 'bad-idea', answerWith: getGithubInfo },
										{ title: 'NO', testId: 'nah', returnValue: false },
									],
								})
								reportAnswer(answer)
							}}
						>
							click?
						</button>
						{answered && <p>Question has been Answered</p>}
					</React.Fragment>
				)}
			</QuestionWrapper>,
		)

		const askButton = getByTestId('askButton')
		/* Act */
		/* open the dialog */
		fireEvent.click(askButton)
		await wait()
		// expect(queryByText('Question has been Answered')).toBeFalsy()

		/* first answer */
		const answerBtn1 = getByTestId('answerBtn1')
		fireEvent.click(answerBtn1)
		await wait()

		// expect(queryByText('Question has been Answered')).toBeTruthy()
		/* open the dialog */
		fireEvent.click(askButton)
		await wait()
		// expect(queryByText('Question has been Answered')).toBeFalsy()

		/* second answer */
		const answerBtn2 = getByTestId('answerBtn2')
		fireEvent.click(answerBtn2)
		await wait()

		/* Assert */
		expect(getGithubInfo.mock.calls).toHaveLength(2)
		expect(getGithubInfo.mock.calls[0][0]).toBe('good-idea')
		expect(getGithubInfo.mock.results[0].value).toBe('https://github.com/good-idea')
		expect(getGithubInfo.mock.calls[1][0]).toBe('bad-idea')
		expect(getGithubInfo.mock.results[1].value).toBe('https://github.com/bad-idea')
	})
})
