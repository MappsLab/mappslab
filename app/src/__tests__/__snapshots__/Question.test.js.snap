// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component Mounts Correctly 1`] = `
<div>
  <button
    data-testid="button"
    type="button"
  >
    click?
  </button>
</div>
`;
